### PR TITLE
Support Result values in Experiments

### PR DESCRIPTION
**Pull request by nate-onesignal**
More often than not, these kind of refactors will involve failable
methods. For this, thesis provides the `run_result` method which adds
special handling, metrics, and fallthrough logic for failable futures.

### CODE REVIEW

**Review by alexgenco — APPROVED** [review]

File: src/experiment.rs [hunk 1]
```diff
@@ -169,6 +165,111 @@ impl<T, C, E, R, M> Experiment<T, C, E, R, M> {
...
+                                    outcome_mismatch(self.name);
+
+                                    let mismatch = Mismatch {
+                                        control: Err(control),
+                                        experimental: Ok(experimental),
+                                    };
+
+                                    return (self.on_mismatch)(mismatch);
+                            }
+                            (Err(control), Err(_)) => {
```
[the remark this answers is not in the document]
Comment (nate-onesignal) [reply src/experiment.rs:261]: I thought that one major push behind this feature is that most Error types do not implement `Eq`

[the remark this answers is not in the document]
Comment (alexgenco) [reply src/experiment.rs:261]: Right, forgot about that. I suppose we would need to support configuring a custom comparator for that, and that probably wouldn't even need to be unique to the `run_result` case. But yeah either way, shouldn't hold up this PR.

[the remark this answers is not in the document]
Comment (nate-onesignal) [reply src/experiment.rs:261]: Yes, in my mind, the `on_mismatch` function serves as that custom comparator, and I didn't think it made sense to call it for the `Err/Err` case.

[the remark this answers is not in the document]
Comment (alexgenco) [reply src/experiment.rs:261]: I'm thinking the custom comparator could be earlier on, basically i.e. https://github.com/OneSignal/thesis/blob/1e9c5251bbd80124653305d87d33b38c2f83e823/src/experiment.rs#L148
you would instead call a custom comparator function that defaulted to `==`. For instance if you're comparing JSON payloads with unique request IDs, you could override the comparator to ignore that field. But again, that's a separate feature.

[the remark this answers is not in the document]
Comment (nate-onesignal) [reply src/experiment.rs:261]: Oh, I see. That's very interesting

**Pull request merged**